pandas/core/reshape/merge.py

```diff
@@ -1866,27 +1873,23 @@ def _factorize_keys(lk, rk, sort=True):
         lk = ensure_int64(lk.codes)
         rk = ensure_int64(rk)
 
-    elif (
-        is_extension_array_dtype(lk.dtype)
-        and is_extension_array_dtype(rk.dtype)
-        and lk.dtype == rk.dtype
-    ):
+    elif is_extension_array_dtype(lk.dtype) and lk.dtype == rk.dtype:
```

Comment:
Is this guaranteed to be equivalent?

Reply:
I'm pretty sure.  How could you have this condition but not `is_extension_array_dtype(rk.dtype)`?

Reply:
should change to lk.is_dtype_equal(rk) (technically the lk.dtype == rk.dtype is unsafe and we should never use this)

Reply:
will update